// @primer/react/lib-esm/ProgressBar/ProgressBar.js
import React__default from 'react';
import styled from 'styled-components';
import { width } from 'styled-system';
import { get } from '../constants.js';
import sx from '../sx.js';

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
const Bar = styled.span.withConfig({
  displayName: "ProgressBar__Bar",
  componentId: "sc-1jz8j7n-0"
})(["width:", ";", ";"], props => props.progress ? `${props.progress}%` : 0, sx);
const sizeMap = {
  small: '5px',
  large: '10px',
  default: '8px'
};
const ProgressContainer = styled.span.withConfig({
  displayName: "ProgressBar__ProgressContainer",
  componentId: "sc-1jz8j7n-1"
})(["display:", ";overflow:hidden;background-color:", ";border-radius:", ";height:", ";", " ", ";"], props => props.inline ? 'inline-flex' : 'flex', get('colors.border.default'), get('radii.1'), props => sizeMap[props.barSize || 'default'], width, sx);
function ProgressBar({
  progress,
  bg = 'success.emphasis',
  barSize = 'default',
  ...rest
}) {
  return /*#__PURE__*/React__default.createElement(ProgressContainer, _extends({
    barSize: barSize
  }, rest), /*#__PURE__*/React__default.createElement(Bar, {
    progress: progress,
    sx: {
      bg
    }
  }));
}
ProgressBar.displayName = "ProgressBar";

export { ProgressBar as default };
